### PR TITLE
RFD 91: Non-AD Desktop Access

### PR DESCRIPTION
**Pull request by probakowski**
Issue: https://github.com/gravitational/teleport/issues/10166

This RFD outlines design for Non-AD Desktop Access

### CODE REVIEW

File: rfd/0091-non-ad-desktop-access.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+- Engineering: @zmb3 && (@ibeckermayer || @LKozlowski)
+- Product: (@klizhentas || @xinding33)
+
+## Overview
+
+Common ask for Desktop Access in Teleport is to connect to hosts that are not part of the domain, so there's no Active
+Directory available and no Kerberos involved. Unfortunately Windows supports smart card login, used currently in
+Teleport, only for domain-joined hosts, so we need another solution for non-AD case.
+
+Idea is to use install on host DLL library containing custom
```
Comment (zmb3) [inline rfd/0091-non-ad-desktop-access.md:19]: I assume (to start at least) that very little of this code overlaps with anything Teleport does today. Does that mean we should create a new repo for this Windows package, or do you plan to include the source in this repository?

File: rfd/0091-non-ad-desktop-access.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+-  Windows Desktop ------|   Winlogon.exe   |   Credential   |                  | 
+|     Service      |     |                  |<---------------|  Verify cert     | 
++------------------+     +------------------+                +------------------+ 
+                            ^  |            +-------------------+     +---------+
+                            |  | Credential |   Authentication  |     |         |           
+                            |  |----------->|      package      |<--->|   SAM   |
+                            |               |                   |     |         |           
+                            |     Token     | Verify credential |     +---------+           
+                            |---------------|Create login token |                             
+                                            +-------------------+
```
Comment (ibeckermayer) [inline rfd/0091-non-ad-desktop-access.md:43]: This diagram is slightly confusing, if I'm understanding it correctly we're looking at something like this (?):

Screen Shot 2022-10-06 at 17 12 28

File: rfd/0091-non-ad-desktop-access.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+### Certificate verification
+
+To verify smart card certificate we need to
+use [wincrypt API](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/). Using it from authentication provider
+is impossible - `lsass.exe` runs in different context and doesn't have access
+to [Microsoft Base Smart Card provider](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/secsmart/microsoft-base-smart-card-cryptographic-service-provider)
+. We have to obtain information to verify that user has correct certificate in credential provider. To do that we will
+sign current time + secret shared between credential provider and authentication package using smart card.
+
+Signature will then be passed along certificate to authentication package, which will verify that both are valid.
```
Comment (ibeckermayer) [inline rfd/0091-non-ad-desktop-access.md:123]: "certificate" here is the same thing (or a part of) the "Credential" mentioned above?

Related to zac's question https://github.com/gravitational/teleport/pull/17134/files#r989341860

File: rfd/0091-non-ad-desktop-access.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+Secret mentioned above will be created during authentication package initialization and saved on disk
+in `C:\Windows\System32`.
```
Comment (LKozlowski) [inline rfd/0091-non-ad-desktop-access.md:126]: Does this shared secret is generated only once or it is regenerated after each Windows Desktop reboot?
Do we need to create it  in the `C:\Windows\System32\`? If so, does the authentication package always has the permissions to create and modify files in the `C:\Windows\System32\` directory?

File: rfd/0091-non-ad-desktop-access.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+### Changes on Teleport side
+
+We will reuse most of our current implementation. Only change will be in configuration- LDAP configuration won't be
+required anymore. If it'll be missing we'll assume non-AD host and use default DNS for name resolution.
+
+In first iteration we'll support on static host definition in configuration (no discovery).
+
+Teleport's CA certificate will have to be imported on Windows host, there'll be no support for automatic CA rotation.
+
```
[the remark this answers is not in the document]
Comment (LKozlowski) [reply rfd/0091-non-ad-desktop-access.md:139]: Adding to this: Do we also need to manually enable the Smart Card service, like in here: https://goteleport.com/docs/desktop-access/getting-started/#enable-the-smart-card-service ?

File: rfd/0091-non-ad-desktop-access.md (hunk 3 above)
Comment (LKozlowski) [inline rfd/0091-non-ad-desktop-access.md:123]: How does we pass that signature to the authentication package?

File: rfd/0091-non-ad-desktop-access.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+- `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Lsa\Authentication Packages`
+
+### Certificate verification
+
+To verify smart card certificate we need to
+use [wincrypt API](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/). Using it from authentication provider
+is impossible - `lsass.exe` runs in different context and doesn't have access
+to [Microsoft Base Smart Card provider](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/secsmart/microsoft-base-smart-card-cryptographic-service-provider)
+. We have to obtain information to verify that user has correct certificate in credential provider. To do that we will
+sign current time + secret shared between credential provider and authentication package using smart card.
```
Comment (LKozlowski) [inline rfd/0091-non-ad-desktop-access.md:121]: Could you describe the signing process in more detail? Are we using the `Microsoft Base Smart Card provider` for the signing purposes?

**Pull request closed without merging**